core/finality-grandpa/src/lib.rs

```diff
@@ -0,0 +1,767 @@
+
+use std::collections::{VecDeque, HashMap};
+use std::sync::Arc;
+use std::time::{Instant, Duration};
+
+const LAST_COMPLETED_KEY: &[u8] = b"grandpa_completed_round";
+
+/// A GRANDPA message for a substrate chain.
+pub type Message<Block> = grandpa::Message<<Block as BlockT>::Hash>;
+/// A signed message.
+pub type SignedMessage<Block> = grandpa::SignedMessage<<Block as BlockT>::Hash, ed25519::Signature, AuthorityId>;
+
+/// Configuration for the GRANDPA service.
+pub struct Config {
+	/// The expected duration for a message to be gossiped across the network.
+	pub gossip_duration: Duration,
+	/// The voters.
+	// TODO: make dynamic
+	pub voters: Vec<AuthorityId>,
+	/// The local signing key.
+	pub local_key: Option<Arc<ed25519::Pair>>,
+}
+
+/// Errors that can occur while voting in GRANDPA.
+#[derive(Debug)]
+pub enum Error {
+	/// An error within grandpa.
+	Grandpa(GrandpaError),
+	/// A network error.
+	Network(String),
+	/// A blockchain error.
+	Blockchain(String),
+	/// A timer failed to fire.
+	Timer(::tokio::timer::Error),
+}
+
+impl From<GrandpaError> for Error {
+	fn from(e: GrandpaError) -> Self {
+		Error::Grandpa(e)
+	}
+}
+
+/// A handle to the network. This is generally implemented by providing some
+/// handle to a gossip service or similar.
+///
+/// Intended to be a lightweight handle such as an `Arc`.
+pub trait Network: Clone {
+	/// A stream of input messages for a topic.
+	type In: Stream<Item=Vec<u8>,Error=()>;
+
+	/// Get a stream of messages for a specific round. This stream should
+	/// never logically conclude.
+	fn messages_for(&self, round: u64) -> Self::In;
+
+	/// Send a message at a specific round out.
+	fn send_message(&self, round: u64, message: Vec<u8>);
+
+	/// Clean up messages for a round.
+	fn drop_messages(&self, round: u64);
+}
+
+/// Something which can determine if a block is known.
+pub trait BlockStatus<Block: BlockT> {
+	/// Return `Ok(Some(number))` or `Ok(None)` depending on whether the block
+	/// is definitely known and has been imported.
+	/// If an unexpected error occurs, return that.
+	fn block_number(&self, hash: Block::Hash) -> Result<Option<u32>, Error>;
+}
+
+impl<B, E, Block: BlockT> BlockStatus<Block> for Arc<Client<B, E, Block>> where
+	B: Backend<Block, Blake2Hasher>,
+	E: CallExecutor<Block, Blake2Hasher>,
+	NumberFor<Block>: As<u32>,
+{
+	fn block_number(&self, hash: Block::Hash) -> Result<Option<u32>, Error> {
+		self.block_number_from_id(&BlockId::Hash(hash))
+			.map_err(|e| Error::Blockchain(format!("{:?}", e)))
+			.map(|num| num.map(|n| n.as_()))
+	}
+}
+
+/// Buffering imported messages until blocks with given hashes are imported.
+pub struct UntilImported<Block: BlockT, Status, I> {
+	import_notifications: Fuse<ImportNotifications<Block>>,
+	status_check: Status,
+	inner: Fuse<I>,
+	ready: VecDeque<SignedMessage<Block>>,
+	check_pending: Interval,
+	pending: HashMap<Block::Hash, Vec<SignedMessage<Block>>>,
```

Comment:
Can this be attacked to grow indefinitely?

Reply:
potentially yes. i'm not sure what the mitigation strategy is there

Reply:
i'll file that as an issue

Reply:
https://github.com/paritytech/substrate/issues/873